definitions/scenarios/self_upgrade.rb

```diff
@@ -58,15 +66,44 @@ class SelfUpgrade < SelfUpgradeBase
       manual_detection
     end
 
-    def compose
+    def downstream_self_upgrade(pkgs_to_update)
       if check_min_version('foreman', '2.5') || check_min_version('foreman-proxy', '2.5')
-        pkgs_to_update = %w[satellite-maintain rubygem-foreman_maintain]
         yum_options = req_repos_to_update_pkgs.map do |id|
           "--enablerepo=#{id}"
         end
         add_step(Procedures::Packages::Update.new(packages: pkgs_to_update, assumeyes: true,
                                                   yum_options: yum_options))
       end
     end
+
+    def upstream_self_upgrade(pkgs_to_update)
+      # This method is responsible for
+      # 1. Setup the repositories of next major version
+      # 2. Update the foreman-maintain packages from next major version repository
+      # 3. Rollback the repository to current major version
+
+      add_step(Procedures::Repositories::Setup.new(:version => upstream_target_version))
+      add_step(Procedures::Packages::Update.new(packages: pkgs_to_update, assumeyes: true))
+    ensure
+      installed_release_pkg = package_manager.find_installed_package('foreman-release',
+                                                                     '%{VERSION}')
+      current_major_version = current_version[0..2]
+      installed_foreman_release_major_version = installed_release_pkg[0..2]
+      if installed_foreman_release_major_version != current_major_version
+        add_step(Procedures::Packages::Uninstall.new(packages: %w[foreman-release katello-repos],
+                                                     assumeyes: true))
+        add_step(Procedures::Repositories::Setup.new(:version => current_major_version))
```

Comment:
I wonder if `Setup` should be able to deal with `foreman-release` and `katello-repos` being present, possibly with a newer version. That would remove the need to uninstall and I mostly say that because if `Uninstall` successed but `Setup` fails you end up with no repositories present. If it's a single call to the package manager then it becomes atomic.

Reply:
Hmm.. I can install the next version or updated release packages without uninstalling the existing one and let the `dnf/yum/apt` do the job. So removing the uninstall step should be a go

Reply:
Yeah so for downgrade I think Uninstall and Setup is correct way. Yeah Setup step takes care of installing the specified version(new/any).